apptools/undo/tests/test_command_stack.py

```diff
@@ -0,0 +1,143 @@
+        with assert_n_commands_pushed(self.stack, 1):
+            self.stack.push(self.command)
+
+    def test_n_command_pushed(self):
+        n = 4
+        with assert_n_commands_pushed(self.stack, n):
+            for i in range(n):
+                self.stack.push(self.command)
+
+    def test_undo_1_command(self):
+        with assert_n_commands_pushed_and_undone(self.stack, 1):
+            self.stack.push(self.command)
+            self.assertEqual(self.stack.undo_name, self.command.name)
+            self.stack.undo()
+
+    def test_undo_n_command(self):
+        n = 4
+        with assert_n_commands_pushed_and_undone(self.stack, n):
+            for i in range(n):
+                self.stack.push(self.command)
+
+            for i in range(n):
+                self.stack.undo()
+
+    def test_undo_unnamed_command(self):
+        unnamed_command = UnnamedCommand()
+        with assert_n_commands_pushed(self.stack, 1):
+            self.stack.push(unnamed_command)
+
+            # But the command cannot be undone because it has no name
+            self.assertEqual(self.stack.undo_name, "")
+            # This is a no-op
+            self.stack.undo()
+
+    def test_undo_redo_1_command(self):
+        with assert_n_commands_pushed(self.stack, 1):
+            self.stack.push(self.command)
+            self.stack.undo()
+            self.stack.redo()
+
+    def test_define_macro(self):
+        with assert_n_commands_pushed(self.stack, 1):
+            add_macro(self.stack, num_commands=2)
+
+    def test_undo_macro(self):
+        with assert_n_commands_pushed_and_undone(self.stack, 1):
+            # The 2 pushes are viewed as 1 command
+            add_macro(self.stack, num_commands=2)
+            self.stack.undo()
+
+    def test_empty_stack_is_clean(self):
+        self.assertTrue(self.stack.clean)
+
+    def test_non_empty_stack_is_dirty(self):
+        self.stack.push(self.command)
+        self.assertFalse(self.stack.clean)
+
+    def test_make_clean(self):
+        # This makes it dirty by default
+        self.stack.push(self.command)
+        # Make the current tip of the stack clean
+        self.stack.clean = True
+        self.assertTrue(self.stack.clean)
+
+    def test_make_dirty(self):
+        # Start from a clean state:
+        self.stack.push(self.command)
+        self.stack.clean = True
+
+        self.stack.clean = False
+        self.assertFalse(self.stack.clean)
+
+
+def add_macro(stack, num_commands=2):
+    command = SimpleCommand()
+    stack.begin_macro('Increment n times')
+    try:
+        for i in range(num_commands):
+            stack.push(command)
+    finally:
+        stack.end_macro()
+
+
+    # Assertion helpers -------------------------------------------------------
+from contextlib import contextmanager
+from nose.tools import assert_equal
+
+@contextmanager
+def assert_n_commands_pushed(stack, n):
+    current_length = len(stack._stack)
+    try:
+        yield
+    finally:
```

Comment:
I think this is one of the cases where a `try/finally` block isn't appropriate for a context manager. I think this will suppress errors in the test and, instead, raise assertion errors.